Add check config.php for ffmpeg_path in PreviewManager.php. fix for snap no video preview

Jachhj-sc opened the pull request:
add a check in config.php for a configured movieBinary path. so now it first checks in config.php if preview_ffmpeg_path is configured. this makes it possible to install the ffmpeg binary and use it with the previewprovider and get video previews in the snap package.

So to get previews in videos with this fix.
1. adding ffmpeg binary to /var/snap/nextcloud/current/[create dir 'bin']/
2. adding the following line to config.php
```
'preview_ffmpeg_path' => '/var/snap/nextcloud/current/bin/ffmpeg',
```
3. sudo snap restart nextcloud

Signed-off-by: William <william.hak57@gmail.com>


* Resolves:  https://github.com/nextcloud-snap/nextcloud-snap/issues/1046

## Summary


## TODO

- [ ]

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

Jachhj-sc committed:
Update PreviewManager.php

add a check in config.php for a configured movieBinary path.
so now it first checks in config.php if  preview_ffmpeg_path is configured.

Signed-off-by: William <william.hak57@gmail.com>

solracsf commented on lib/private/PreviewManager.php:
```diff
@@ -417,11 +417,16 @@ protected function registerCoreProviders() {
-			$movieBinary = $this->binaryFinder->findBinaryPath('avconv');
-			if (!is_string($movieBinary)) {
-				$movieBinary = $this->binaryFinder->findBinaryPath('ffmpeg');
+
+			$movieBinary = $this->config->getSystemValue('preview_ffmpeg_path', null);
+			if(!is_string($movieBinary)){
```
```suggestion
			if (!is_string($movieBinary)) {
```

szaimen replied:
Should fix lint:
```suggestion
			$movieBinary = $this->config->getSystemValue('preview_ffmpeg_path', null);
			if (!is_string($movieBinary)){
```

szaimen commented:
Hi, thanks a lot for your PR! Please add this value also to https://github.com/nextcloud/server/blob/master/config/config.sample.php. in order to document it. Thank you! :)

tcitworld commented:
I'm not sure why the snap package can't just set `/var/snap/nextcloud/current/bin` (or whatever it looks like from inside) into the `$PATH` so that Nextcloud can use binaries put there straight away?

szaimen commented:
> I'm not sure why the snap package can't just set `/var/snap/nextcloud/current/bin` (or whatever it looks like from inside) into the `$PATH` so that Nextcloud can use binaries put there straight away?

cc @kyrofa ?

kyrofa commented:
Because there's nothing useful there: the snap doesn't include ffmpeg, and it's read-only so users can't put things there either. However, users CAN download static ffmpeg binaries and put them somewhere the snap _can_ access, they just need to be able to tell Nextcloud where they are.

szaimen commented:
Yeah, fine to add it imho.

@nextcloud/server-backend WDYT?

nickvergessen commented:
Config is fine, but should be documented in the sample file

tcitworld commented:
In that case, instead of adding a specific config key for the ffmpeg binary (for instance, [the memories app also needs `ffprobe`](https://github.com/nextcloud-snap/nextcloud-snap/issues/1046#issuecomment-1374427458)), let's add a generic `extra_path` config key that can be used for the `OC/BinaryFinder`  class.

szaimen commented:
> Config is fine, but should be documented in the sample file

Thanks Joas! Then @Jachhj-sc please do as advised :)

Jachhj-sc commented:
alright ill document it in the sample file.
shall i create a separate pull request, or link to this one for that? or what is the proper way?

Jachhj-sc committed:
add documentation preview_ffmpeg_path

add preview_ffmpeg_path documentation.
document custom path for ffmpeg so it can be used by the previewprovider to create video previews with the snap package of nextcloud.

Signed-off-by: William <william.hak57@gmail.com>

Jachhj-sc committed:
Merge branch 'nextcloud:master' into patch-1

Jachhj-sc committed:
Merge pull request #1 from Jachhj-sc/patch-2

add documentation preview_ffmpeg_path

szaimen reviewed (approved):
LGTM but didnt test

nickvergessen commented on config/config.sample.php:
```diff
@@ -1190,7 +1190,14 @@
+/**
+ * custom path for ffmpeg binary
+ * 
+ * Defaults to ``''`` (empty string)
```
```suggestion
 * Defaults to ``null`` and falls back to searching ``avconv`` and ``ffmpeg`` in the configured ``PATH`` environment
```

---

Can also default to empty string as documented, but then you should use `getSystemValueString` in your code change
```php
$movieBinary = $this->config->getSystemValueString('preview_ffmpeg_path');
```

Jachhj-sc committed:
Update config/config.sample.php

add better description

Co-authored-by: Joas Schilling <213943+nickvergessen@users.noreply.github.com>
Signed-off-by: William <william.hak57@gmail.com>

Jachhj-sc committed:
Update lib/private/PreviewManager.php

fix lint issue

Co-authored-by: Simon L. <szaimen@e.mail.de>
Signed-off-by: William <william.hak57@gmail.com>

szaimen commented:
> In that case, instead of adding a specific config key for the ffmpeg binary (for instance, [the memories app also needs `ffprobe`](https://github.com/nextcloud-snap/nextcloud-snap/issues/1046#issuecomment-1374427458)), let's add a generic `extra_path` config key that can be used into the `OC/BinaryFinder` class. Then the calls to `findBinaryPath` will take this new path into account.

Oh sorry, I overread this. Not sure about this apporach. Sounds a bit over-engineered to me...

szaimen commented:
Wdyt @nickvergessen?

nickvergessen commented:
yeah, also not sure about that.

szaimen commented:
Okay, lets merge then as is since the memories app already has config values anyway :)

szaimen commented:
CI failure unrelated

ZOKOB commented:
Hello I followed your tutorial however I still don't have my video thumbnails here is my configuration file

  'memories.exiftool' => '/var/snap/nextcloud/33730/nextcloud/extra-apps/memories/exiftool-bin/exiftool-am
  'memories.vod.ffmpeg' => '/var/snap/nextcloud/bin/ffmpeg',
  'memories.vod.ffprobe' => '/var/snap/nextcloud/bin/ffprobe',
  'memories.vod.disable' => false,
  'memories.ffmpeg_path' => '/snap/ffmpeg-integration-gplv3/1/bin/ffmpeg',
  'memories.ffprobe_path' => '/snap/ffmpeg-integration-gplv3/1/bin/ffprobe',
  'memories.tmp_path' => '/var/tmp',
  'memories.vod.path' => '/var/snap/nextcloud/33908/nextcloud/extra-apps/memories/exiftool-bin/go-vod-amd64',
  'preview_ffmpeg_path' => '/var/snap/nextcloud/bin/ffmpeg',
  'enable_previews' => true,
  'enabledPreviewProviders' =>
   array (
    0 => 'OC\Preview\TXT',
    1 => 'OC\Preview\MarkDown',
    2 => 'OC\Preview\OpenDocument',
    3 => 'OC\Preview\PDF',
    4 => 'OC\Preview\MSOffice2003',
    5 => 'OC\Preview\MSOfficeDoc',
    6 => 'OC\Preview\Image',
    7 => 'OC\Preview\Photoshop',
    8 => 'OC\Preview\TIFF',
    9 => 'OC\Preview\SVG',                                                            10 => 'OC\Preview8\Font',
   11 => 'OC\Preview\MP3',
   12 => 'OC\Preview\Movie',
   13 => 'OC\Preview\MKV',
   14 => 'OC\Preview\MP4',
   15 => 'OC\Preview\AVI',
   )
);`
can you help me please?

szaimen commented:
Are you on Nextcloud 26?

ZOKOB commented:
how to update via the snap tool in 26?

szaimen commented:
Please ask here: https://github.com/nextcloud-snap/nextcloud-snap/issues/2348
